src/RestApi/StoreApi/Utilities/CartController.php

```diff
@@ -156,6 +156,80 @@ public function empty_cart() {
 		wc()->cart->empty_cart();
 	}
 
+	/**
+	 * See if cart has applied coupon by code.
+	 *
+	 * @param string $coupon_code Cart coupon code.
+	 * @return bool
+	 */
+	public function has_coupon( $coupon_code ) {
+		return wc()->cart->has_discount( $coupon_code );
+	}
+
+	/**
+	 * Returns all applied coupons.
+	 *
+	 * @param callable $callback Optional callback to apply to the array filter.
+	 * @return array
+	 */
+	public function get_cart_coupons( $callback = null ) {
+		return $callback ? array_filter( wc()->cart->get_applied_coupons(), $callback ) : array_filter( wc()->cart->get_applied_coupons() );
+	}
+
+	/**
+	 * Based on the core cart class but returns errors rather than rendering notices directly.
+	 *
+	 * @throws RestException Exception if invalid data is detected.
+	 * @param string $coupon_code Coupon code.
+	 */
+	public function apply_coupon( $coupon_code ) {
```

Comment:
Definitely but there isn't much choice without refactoring core. The problem with that method, like the add to cart one, is that they enqueue notices. This results in a spam of notices showing up on store pages after running a few API requests. 

Hopefully we have time to work on some refactors as these APIs reveal problems such as this one. Right now however I'm just focussed on getting working endpoints.

Reply:
> Right now however I'm just focussed on getting working endpoints.

Works for me.  Should we track these things as we notice them so it can potentially result in issues/work slated for core at some point?

Reply:
I wonder if one way we could track these things is to introduce a `known technical debt` label that we can assign to pulls that introduce _known_ tech debt that we need to address at some point.  Then we can also include a section in the pull description about what specific technical debt is being introduced?

Then at some future date, we have a way to retrieve all pulls with this label to prioritize knocking off the tech-debt (and remove the label with a new issue or pull)?

Reply:
I was going to add some inline `@todo` items so I can add a little more description in context. Looks like there are some in the codebase already, and some outdated ones so I'll skim through at the same time.

Reply:
I think inline @todo's are good because they can easily be surfaced in our IDE's... however, we still need to remember to review them intentionally otherwise they are still easily forgotten.  Do you have suggestions on what would be the best way to surface those todos so they aren't forgotten?

(BTW, none of this convo is blocking for merging this pull, I just want to ensure we don't forget about this tech debt or it gets shelved indefinitely).

Reply:
@nerrad Got it. I'm working on the price responses before completing this pull since there are some totals rows here too which will need adjustment.